### PR TITLE
[SPARK-35373][BUILD] Check Maven artifact checksum in build/mvn

### PR DESCRIPTION
**Pull request by srowen**
### What changes were proposed in this pull request?

`./build/mvn` now downloads the .sha512 checksum of Maven artifacts it downloads, and checks the checksum after download.

### Why are the changes needed?

This ensures the integrity of the Maven artifact during a user's build, which may come from several non-ASF mirrors.

### Does this PR introduce _any_ user-facing change?

Should not affect anything about Spark per se, just the build.

### How was this patch tested?

Manual testing wherein I forced Maven/Scala download, verified checksums are downloaded and checked, and verified it fails on error with a corrupted checksum.

### CODE REVIEW

**Review by dongjoon-hyun — APPROVED** [review]
Looks good to me.

In case of the mentioned out-of-sync cases between Apache main and mirrors, we will hit some random Maven jobs failures after this PR. It's rare and hopefully it will be recovered automatically  in the subsequent runs eventually.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by srowen** [comment]
I may merge this to master first to see if there are issues, then back-port it if it seems OK. We probably do want this as part of maintenance branches.

**Comment by dongjoon-hyun** [comment]
+1 for your decision, @srowen . Ya. Let's merge to master first and monitor it.

**Comment by srowen** [comment]
Merged to master. I'll back port in a while if there are no issues.

**Comment by srowen** [comment]
Backported to 3.1 / 3.0. There were merge conflicts in 2.4 ; maybe not worth backporting there as it's EOL

**Comment by dongjoon-hyun** [comment]
Oh, @srowen .
This break `branch-3.1`.
- https://github.com/apache/spark/runs/2620955249
```
exec: curl --silent --show-error -L https://downloads.lightbend.com/zinc/0.3.15/zinc-0.3.15.tgz/zinc-0.3.15.tgzzinc-0.3.15/bin/zinc
./build/mvn: line 56: /home/runner/work/spark/spark/build/: Is a directory
exec: wget --no-verbose https://downloads.lightbend.com/zinc/0.3.15/zinc-0.3.15.tgz/zinc-0.3.15.tgzzinc-0.3.15/bin/zinc
/home/runner/work/spark/spark/build/: Is a directory
ERROR: Cannot download https://downloads.lightbend.com/zinc/0.3.15/zinc-0.3.15.tgz/zinc-0.3.15.tgzzinc-0.3.15/bin/zinc with cURL or wget; please install manually and try again.
Error: Process completed with exit code 2.
```

**Comment by dongjoon-hyun** [comment]
It seems that the root cause is `Zinc` is removed at master branch while it exists in old branches.

**Comment by dongjoon-hyun** [comment]
I made a follow-up.
- https://github.com/apache/spark/pull/32591

File: build/mvn
Comment (maropu) [inline build/mvn:77]: We don't need to add `$(command -v shasum)` here to check if  `shasum` exists or not (then, raise an error with an explicit message)? For example, Amazon Linux2 does not have it by default.

Comment (srowen) [reply build/mvn:77]: Oh really? shoot. That seemed to be the one command that was on Linux and OS X. Well, we can fail outright if not present, which would stop the build. Is that better or worse than just silently letting it proceed without verification? I don't know. Would developers generally build on this distro?

Comment (maropu) [reply build/mvn:77]: > Oh really? shoot. That seemed to be the one command that was on Linux and OS X.

Yea, for Amazon Linux2, we need to run `sudo yum install perl-Digest-SHA` before running `build/mvn`.

> Is that better or worse than just silently letting it proceed without verification?

IMO, since this check exists to prevent cyber attacks, it is safer to make`build/mvn` fail if  `shasum` does not exists (actually, it seems the other OSs have `shasum` by default, so most developers don't care, I think).

> Would developers generally build on this distro?

Not sure, but that's the default option in aws ec2. I noticed this issue because my daily build broke after this commit merged.

Comment (srowen) [reply build/mvn:77]: OK. I can submit a follow-up to fail if shasum is not present. That might inconvenience some developers, but, this is a security issue, and of course developers are welcome to install mvn / scala directly if desired.

Comment (maropu) [reply build/mvn:77]: +1

File: build/mvn
Comment (Yikun) [reply build/mvn:54]: `build/mvn: line 56: [: /root/spark/build/apache-maven-3.6.3-bin.tar.gz: binary operator expected` is raised after this patch, I submitted a followup:

https://github.com/apache/spark/pull/32608

**Pull request closed without merging**
